src/main/java/org/springframework/data/redis/listener/RedisMessageListenerContainer.java

```diff
@@ -689,36 +690,55 @@ private class SubscriptionTask implements SchedulingAwareRunnable {
 		private class PatternSubscriptionTask implements SchedulingAwareRunnable {
 
 			private long WAIT = 500;
-			private long ROUNDS = 3;
+			private AtomicBoolean isThreadRunning = new AtomicBoolean(true);
+
 
 			public boolean isLongLived() {
 				return false;
 			}
 
+			void cancel() {
+				isThreadRunning.set(false);
+			}
+
 			public void run() {
 				// wait for subscription to be initialized
 				boolean done = false;
 				// wait 3 rounds for subscription to be initialized
-				for (int i = 0; i < ROUNDS && !done; i++) {
+				while(!done && isThreadRunning.get() && !Thread.currentThread().isInterrupted()) {
 					if (connection != null) {
-						synchronized (localMonitor) {
-							if (connection.isSubscribed()) {
-								done = true;
-								connection.getSubscription().pSubscribe(unwrap(patternMapping.keySet()));
-							} else {
-								try {
-									Thread.sleep(WAIT);
-								} catch (InterruptedException ex) {
-									Thread.currentThread().interrupt();
+							try {
+								if (connection.isSubscribed()) {
+									synchronized (localMonitor) {
```

Comment:
this is synchronized by `localMonitor` to make sure the thread exit cleanly when canceled by the `SubscriptionTask`. This should makes the shutdown more graceful and ordered if the container is shutdown while booting and subscription to patterns